// appium-device-farm/lib/test/e2e/android/conf3.spec.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const webdriverio_1 = require("webdriverio");
const APPIUM_HOST = '127.0.0.1';
const APPIUM_PORT = 31337;
const WDIO_PARAMS = {
    connectionRetryCount: 0,
    hostname: APPIUM_HOST,
    port: APPIUM_PORT,
    path: '/wd/hub/',
    logLevel: 'info',
};
const capabilities = {
    platformName: 'Android',
    'appium:uiautomator2ServerInstallTimeout': '50000',
    'appium:automationName': 'UIAutomator2',
    'appium:app': 'https://github.com/AppiumTestDistribution/appium-demo/blob/main/VodQA.apk?raw=true',
};
describe('Plugin1 Test', () => {
    let driver;
    beforeEach(() => __awaiter(void 0, void 0, void 0, function* () {
        driver = yield (0, webdriverio_1.remote)(Object.assign(Object.assign({}, WDIO_PARAMS), { capabilities }));
    }));
    it('Slider test', () => __awaiter(void 0, void 0, void 0, function* () {
        console.log(yield driver.capabilities.deviceUDID);
        yield driver.$('~login').click();
        yield driver.$('~slider1').click();
    }));
    afterEach(() => __awaiter(void 0, void 0, void 0, function* () {
        if (driver) {
            yield driver.deleteSession();
            driver = null;
        }
    }));
});
